# Extracting Data from Foursquare
06/5/19

* Refactor code from "Exploring Foursquare"

# Libraries and Functions

In [2]:
print("Importing ...")
import numpy as np # library to handle data in a vectorized manner
# import math   # for ceil function

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

# #!conda install -c conda-forge geopy --yes # uncomment this line if needed
# from geopy.geocoders import Nominatim # convert address into latitude and longitude

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
# import matplotlib.cm as cm
# import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.8.3 --yes # uncomment this line if needed
import folium # map rendering library
from folium import plugins

print("Folium version: ", folium.__version__)

Importing ...
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.8.3


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.8.3               |             py_0          54 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    openssl-1.1.1b             |       h14c3975_1         4.0 MB  conda-forge
    altair-3.1.0               |           py36_0         724 KB  conda-forge
    ca-certificates-2019.6.16  |       hecc5488_0         145 KB  conda-forge
    certifi-2019.6.16          |           py36_0         148 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         5.1 MB

The following NEW pac

In [3]:
# @hidden cell
# Foursquare information

CLIENT_ID = 'CWCPNAVXDH3TI1BGS4VED4ANSUKEFGHBA4511GRPYPKPNJRD' # your Foursquare ID
CLIENT_SECRET = 'ALMFEYIENSPH3RV3TQB1NGWTKTANVJ5QTHAVZ5B1GRJWIP21' # your Foursquare Secret

In [4]:
# Constants

# Foursquare query information that will be constant during this project
VERSION = '20190214' # Foursquare API version
LIMIT = 120 

# Used to convert latitude and longitude to distances in miles
MILES_PER_LAT = 69    # A degree of latitude is 69 miles. 
MILES_PER_LNG = 50.5  # In Georgia, a degree of longitude averages 50.5 miles

In [5]:
# define function that extracts the category of the venue
# This function comes from Coursera Applied Data Science Capstone class
# used in Foursquare section

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
          
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [6]:
def split_box(box, purpose="number"):
    # Breaks a map "box" into smaller pieces
    # purpose = "size" - if the size of the box exceeds Foursquare's max size
    # purpose = "number" - if the number of returned results is 100 or more,
    #                      indicating that the query "maxed out"
    
    '''Foursquare limits bounding box requests to no more than 10,000 square 
    kilometers, or 3860 square miles. This is a box a bit larger than 
    60 miles on a side.'''
    
    MAX_SIZE_LENGTH = 60

    MAX_DELTA_LAT = MAX_SIZE_LENGTH/MILES_PER_LAT   # Max difference in latitude 
    MAX_DELTA_LNG = MAX_SIZE_LENGTH/MILES_PER_LNG   # Max difference in longitude 
    
    delta_lat = abs(box[0][0] - box[1][0])
    delta_lng = abs(box[0][1] - box[1][1])
    
    boxes = [] # Will hold 1 or more map boxes
    
    if purpose == "number":  # Splitting box because too many results returned
        lat_divisions = lng_divisions = 2 # Break the box into 4 sub-boxesx
        
    elif purpose == "size":  # Splitting because box is too large for Foursquare
        lat_divisions = int(np.ceil(delta_lat / MAX_DELTA_LAT))
        lng_divisions = int(np.ceil(delta_lng / MAX_DELTA_LNG))
        
    else:  
        raise ValueError("This function only supports 'size' and 'number' box splitting")
    
    # Create sub-boxes based on number of lat & lng divisions
    for i in range(lat_divisions):
        for j in range(lng_divisions):
            lat_side = delta_lat / lat_divisions
            lng_side = delta_lng / lng_divisions
            swij = (box[0][0] + i * lat_side, box[0][1] + j * lng_side)
            neij = (swij[0] + lat_side, swij[1] + lng_side)
            boxes.append((swij, neij))            
    
    return boxes   

In [7]:
def query_box(box, codes):
    print(".", end="") # Indicate activity while performing queries

  # define the Foursquare URL

  # Foursquare category ID requires comma-separated text string
    code_string = ",".join(codes)    
    
    url = f'https://api.foursquare.com/v2/venues/explore?&client_id={CLIENT_ID}' + \
        f'&client_secret={CLIENT_SECRET}&v={VERSION}&sw={box[0][0]},{box[0][1]}' + \
        f'&ne={box[1][0]},{box[1][1]}' + \
        f'&categoryId={code_string}&limit={LIMIT}'
  
  # submit the url and capture the returned text
    venues = requests.get(url).json()["response"]['groups'][0]['items']
  
  # if any venues are captured, process and return dataframe. Otherwise, return None. 
    if venues:
        venues = json_normalize(venues) # flatten JSON

        # select desired columns
        filtered_columns = ['venue.name', 'venue.categories', # 'venue.location.postalCode', 
                        'venue.location.lat', 'venue.location.lng', 'venue.location.state']
        venues = venues.loc[:, filtered_columns]

        # filter the category for each row
        venues['venue.categories'] = venues.apply(get_category_type, axis=1)

        # clean column names
        venues.columns = [col.split(".")[-1] for col in venues.columns]
  
        return venues
  
    else:
        return None  # no venues in box meet criteria    

In [8]:
def shrink_box(box, codes, venue_list, query_list):
    
    # Get venues for box (if any) in Pandas dataframe
    box_results = query_box(box, codes)

    try: # This will succeed if 1 or more venues are returned in box_results
        num_clients = box_results.shape[0]

        if num_clients < 100:  
            # Case where results did not "max out" at 100 returned venues
            venue_list += box_results.values.tolist()
            query_list.append((box, num_clients))

        else:  # This is the case where num_clients > 100
            for sub_box in split_box(box, purpose="number"):
                shrink_box(sub_box, codes, venue_list, query_list)

    except: # happens if query returns no results
        query_list.append((box, 0))

    return # venue_list and query_list are mutable, no need to return

In [9]:
def GetVenuesByBox(start_box,  # a tuple containing sw & ne corners of box
                   codes):     # a list of Foursquare category strings

    venue_list = []      # list of venus of interest
    query_list = []      # boxes and number of venues in each, for visualization
    
    # if the start box is too large, split it into smaller boxes
    for box in split_box(start_box, purpose="size"):
        # List will contain one or more boxes. Obtain venues for each
        shrink_box(box, codes, venue_list, query_list)
        
    # Add venue data to a dataframe    
    venues = pd.DataFrame(venue_list, columns=['name', 'categories', # 'postalCode', \
                                                  'lat', 'lng', 'state'])
    return venues, query_list

# Requesting Data from Foursquare

Foursquare allows you to request data by defining a bounding box using latitude & longitude coordinates. I want to request all Asian (including Indian & Pakistani) restaurants in a box drawn around the state of Georgia. 

There are two problems. First, Foursquare has a size limit (about 10,000 square kilometers). Georgia is much larger. Also, Foursquare will only return 100 venues per request (at least with my account type). My code handles these issues by dividing the box into smaller sub-boxes until all boxes are small enough for Foursquare. Then every time a query on a sub-box returns 100 results, the sub-box is split into 4 pieces and a new query is run on each one. This process repeats until all queries return <100 results. Any queries with 100 results are thrown away to avoid duplication.

Let's begin by drawing a large box around Georgia and requesting data. We will cover parts of the adjacent states but we will drop any rows that aren't in Georgia later.

In [10]:
# Define the southwest and northeast corners of a box around the state of Georgia

fsw = (30.357851, -85.605165)  # Southwest corner, Georgia USA
fne = (35.000659, -80.839729)  # Northeast corner

In [11]:
# The following venue categories come from Foursquare. Note that each 
# category has sub-categories (for example, "Korean Restaurant" in "Asian").
restaurant_codes =  ['4bf58dd8d48988d142941735'   # Asian Restaurant
                    ]   

# Request the data
df_venues, query_list = GetVenuesByBox((fsw,fne), restaurant_codes)

......................................................................................................................................................

## Data gathering process

Let's see how the code split the data request. Each box in the picture below represents a Foursquare query that returned 99 or fewer results. The shade of the box shows the number of returned values (you can click on the box to see the number). This is NOT a picture of the data -- we will see that in much more detail later.

In [12]:
# Center the map on the middle of the bounding box defined above 
center = [(fsw[0]+fne[0])/2, (fsw[1]+fne[1])/2]

# Find the maximum number of venues in any square (to calculate fill opacity)
max_clients = max([query[1] for query in query_list])

# Visualize all queries on a map, with fill color indicating number of restaurants
map_georgia = folium.Map(location=center, zoom_start=7, control_scale=True)

plugins.ScrollZoomToggler().add_to(map_georgia)
    
for box in query_list:
    sw, ne = box[0][0], box[0][1]
    opacity = box[1]/max_clients   # Clients in this box / max across all boxes          
    folium.Rectangle([sw, ne], 
                     popup=str(box[1]), color="black", opacity=0.99, fill=True, 
                     fill_color='orange', weight = 1, 
                     fill_opacity=(opacity)).add_to(map_georgia)
    
map_georgia

In the picture above you can see we started with a 6x6 grid, for 36 queries, to satisfy the size limit. After throwing away all queries returning 100 results (the max limit), how many successful queries did we end up with?

In [13]:
# Number of queries returning <100 results
len(query_list)

120

To get to a total of xxx, boxes were split (120 - 36)/3 = 28 times. Data from those boxes were  thrown away, meaning 28 queries were wasted. The total number of queries to get my data from Georgia was 120 +  28 = 148. 

However, the smallest boxes are about 2 square miles in size. Georgia covers almost 60,000 square miles. **Over 25,000 queries would be needed to get this data with equal-sized boxes.** So this method is a bit better than that.

It generally takes 60-90 seconds to get the data in this way.

# Data Cleaning 

Now let's look at the data we've obtained, and get rid of any venues we don't need for our analysis

In [14]:
print("Number of restaurants in data set: ",df_venues.shape[0],"\n\n")
df_venues.head()

Number of restaurants in data set:  4033 




name          categories        lat        lng  \
0  Super Canton Chinese Restaurant  Chinese Restaurant  30.951445 -85.516435   
1                   Fortune Cookie  Chinese Restaurant  30.783262 -85.248280   
2      New Star Chinese Restaurant    Asian Restaurant  30.754914 -85.549341   
3                       King House    Asian Restaurant  30.442998 -85.054370   
4                     Panda Buffet  Chinese Restaurant  30.774442 -85.223421   

  state  
0    FL  
1    FL  
2    FL  
3    FL  
4    FL

My bounding-box approach obtained restaurants in parts of adjacent states. Let's get rid of those.

In [15]:
# Which states are included in the data set?

print(list(set(df_venues.state)))

['Alabama', 'TN', 'GA', 'SC', 'Florida', 'FL', 'South Carolina', 'Georgia', 'AL']


In [16]:
# Get rid of restaurants in other states and count the results.
df_georgia = df_venues[df_venues['state'].isin(['GA', 'Georgia'])]
print("Number of restaurants Georgia: ",df_georgia.shape[0],"\n\n")

Number of restaurants Georgia:  3115 




That's better. In theory, this approach would prevent duplication in the data set, but I always like to be sure. Let's check for duplicate venues.

In [17]:
print("Before dropping duplicates: ",df_georgia.shape[0])
df_georgia = df_georgia.drop_duplicates(keep='first')
print("After dropping duplicates: ",df_georgia.shape[0])

Before dropping duplicates:  3115
After dropping duplicates:  3115


Foursquare did a good job of giving us a unique data set. I also looked at the data and confirmed that all rows contain values for restaurant name, categories, lat, and lng. Now let's look at the restaurant categories. Did we only get the ones we wanted?

In [18]:
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name   lat   lng  state
categories                                        
Chinese Restaurant         1068  1068  1068   1068
Asian Restaurant            666   666   666    666
Japanese Restaurant         455   455   455    455
Sushi Restaurant            246   246   246    246
Korean Restaurant           213   213   213    213
Thai Restaurant             178   178   178    178
Vietnamese Restaurant        93    93    93     93
Noodle House                 32    32    32     32
Ramen Restaurant             14    14    14     14
Bakery                        9     9     9      9
Szechuan Restaurant           9     9     9      9
Restaurant                    8     8     8      8
Seafood Restaurant            8     8     8      8
American Restaurant           7     7     7      7
Taiwanese Restaurant          7     7     7      7
Fast Food Restaurant          6     6     6      6
Food Court                    6     6     6      6
Malay Restaurant              5     5     5      5
Café                          5     5     5      5
Grocery Store                 4     4     4      4
Shabu-Shabu Restaurant        4     4     4      4
Supermarket                   4     4     4      4
Hotpot Restaurant             4     4     4      4
Indonesian Restaurant         4     4     4      4
Dim Sum Restaurant            4     4     4      4
Mongolian Restaurant          3     3     3      3
BBQ Joint                     3     3     3      3
Cantonese Restaurant          3     3     3      3
Poke Place                    3     3     3      3
Indian Chinese Restaurant     3     3     3      3
Coffee Shop                   2     2     2      2
Dumpling Restaurant           2     2     2      2
Steakhouse                    2     2     2      2
Karaoke Bar                   2     2     2      2
Japanese Curry Restaurant     2     2     2      2
Food Truck                    2     2     2      2
Indian Restaurant             2     2     2      2
Sandwich Place                2     2     2      2
Bar                           2     2     2      2
Sports Bar                    1     1     1      1
Gas Station                   1     1     1      1
Bubble Tea Shop               1     1     1      1
Buffet                        1     1     1      1
Shanghai Restaurant           1     1     1      1
Taco Place                    1     1     1      1
Breakfast Spot                1     1     1      1
Tea Room                      1     1     1      1
Soup Place                    1     1     1      1
Deli / Bodega                 1     1     1      1
Caribbean Restaurant          1     1     1      1
Cocktail Bar                  1     1     1      1
Salon / Barbershop            1     1     1      1
Hot Dog Joint                 1     1     1      1
Middle Eastern Restaurant     1     1     1      1
Mexican Restaurant            1     1     1      1
Filipino Restaurant           1     1     1      1
Jiangsu Restaurant            1     1     1      1
Fried Chicken Joint           1     1     1      1
Hunan Restaurant              1     1     1      1
Garden                        1     1     1      1
Hotel Bar                     1     1     1      1
Wings Joint                   1     1     1      1

Convenience store? Italian restaurant? Garden? American restaurant? Clearly we need to get rid of some of these venues. In the cell below, I go through the list of categories and define a list of ones we want to get rid of. I generated the list using this command:

    sorted(list(set(df_georgia['categories'])))

In many cases, I used a command like the following to see which restaurants were in a suspicious category: 

     df_georgia[df_georgia['categories'].isin(['Taco Place'])]

You can see my comments in the lines below.

In [19]:
bad_list = ['American Restaurant',
# 'Asian Restaurant',
# 'BBQ Joint',  # Three venues, mixed Asian
 'Bakery',
 'Bar',
 'Breakfast Spot',
# 'Bubble Tea Shop',
# 'Buffet',   # One, Japanese
 'Café',
# 'Cantonese Restaurant',
 'Caribbean Restaurant',
# 'Chinese Restaurant',
 'Cocktail Bar',
 'Coffee Shop',
 'Deli / Bodega',
# 'Dim Sum Restaurant',
 'Dumpling Restaurant',  # Two restaurants, both Korean, both permanently closed
 'Fast Food Restaurant',
# 'Filipino Restaurant',
 'Food Court',
# 'Food Truck',
# 'Fried Chicken Joint',  # One, Korean
# 'Garden',   # One instance. Chinese restaurant.
 'Gas Station',
 'Grocery Store',
# 'Hot Dog Joint',  # One instance, Korean fried hotdogs. Must. Try. This.
 'Hotel Bar',
# 'Hotpot Restaurant',
# 'Indian Chinese Restaurant',
# 'Indian Restaurant',
# 'Indonesian Restaurant',  # Four instances. Roll into "Asian"
# 'Japanese Curry Restaurant',
# 'Japanese Restaurant',
# 'Jiangsu Restaurant',
# 'Karaoke Bar',
# 'Korean Restaurant',
# 'Malay Restaurant',  # Five instances. Roll into "Asian"
 'Mexican Restaurant',
 'Middle Eastern Restaurant',
# 'Mongolian Restaurant',  Roll into "Asian"
# 'Noodle House',
# 'Poke Place',  # Hawaiian, but we'll call it Japanese
# 'Ramen Restaurant',
 'Restaurant',
 'Salon / Barbershop',
# 'Sandwich Place',  # Two. One is Korean, one is not Asian
 'Seafood Restaurant',
# 'Shabu-Shabu Restaurant',
# 'Shanghai Restaurant',
 'Soup Place',   # One instance, permanently closed
 'Sports Bar',
# 'Steakhouse',  # Two. Both Japanese
# 'Supermarket',  # Four. All have restaurants inside
# 'Sushi Restaurant',
# 'Szechuan Restaurant',
# 'Taco Place', # One location "Hankook Taqueria" Korean
# 'Taiwanese Restaurant',
# 'Tea Room',  # One Location, Chinese
# 'Thai Restaurant',
# 'Vietnamese Restaurant',
 'Wings Joint']


Let's see how many restaurants are in the categories we don't want, and drop them from the data frame.

In [20]:
print("Number of restaurants in unwanted categories (dropped): ", df_georgia[df_georgia['categories'].isin(bad_list)].shape[0])
df_georgia = df_georgia[~df_georgia['categories'].isin(bad_list)]
print("Restaurants remaining in dataset: ", df_georgia.shape[0])

Number of restaurants in unwanted categories (dropped):  71
Restaurants remaining in dataset:  3044


Now let's consolidate some categories. Some of the categories remaining in our database (such as Szechuan Restaurant) are sub-categories. For our purposes, we want to roll these up to the top-level category (Szechuan --> Chinese, for example). Other categories roll up because I looked at the restaurants in the category and found that they all should have been put in another category (for example, all Steakhouses in our database are Japanese). Now I will define lists that will help consolidate categories in these cases.

In [21]:
# Define sub-categories that will be consolidated into major categories.
roll_up = [[['Bubble Tea Shop','Cantonese Restaurant','Dim Sum Restaurant','Garden', \
            'Shanghai Restaurant','Szechuan Restaurant', 'Taiwanese Restaurant', \
            'Tea Room', 'Hunan Restaurant'], 'Chinese Restaurant'],
           [['Buffet','Japanese Curry Restaurant', 'Japanese Restaurant', \
            'Ramen Restaurant', 'Shabu-Shabu Restaurant','Steakhouse', \
            'Sushi Restaurant'], 'Japanese Restaurant'],
           [['Fried Chicken Joint','Hot Dog Joint', 'Taco Place'], 'Korean Restaurant'],
           [['Indian Chinese Restaurant','Indian Restaurant','Jiangsu Restaurant'], \
            'Indo-Pak Restaurant'],
           [['BBQ Joint', 'Filipino Restaurant','Food Truck','Hotpot Restaurant',\
            'Indonesian Restaurant','Karaoke Bar','Malay Restaurant', \
            'Mongolian Restaurant', 'Noodle House','Poke Place', \
            'Sandwich Place', 'Supermarket'], 'Asian Restaurant']]

In [22]:
# Consolidate categories

for cat in roll_up:
    df_georgia['categories'][df_georgia['categories'].isin(cat[0])] = cat[1]

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


In [23]:
# Let's see how the restaurants are distributed now
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name   lat   lng  state
categories                                    
Chinese Restaurant     1096  1096  1096   1096
Asian Restaurant        731   731   731    731
Japanese Restaurant     724   724   724    724
Korean Restaurant       216   216   216    216
Thai Restaurant         178   178   178    178
Vietnamese Restaurant    93    93    93     93
Indo-Pak Restaurant       6     6     6      6

## South Asian - Data Gathering & Cleaning
So far, we've only uncovered a handful of South Asian (Indian & Pakistani) restaurants. That's because Foursquare has separate categories for them. The restaurant supply company also serves Indian & Pakistani restaurants. Let's repeat the above analysis and merge the data. 

In [24]:
# The following venue categories come from Foursquare. Note some 
# categories have sub-categories (for example, "Korean Restaurant" in "Asian").
restaurant_codes =  ['4bf58dd8d48988d10f941735',   # Indian Restaurant
                     '52e81612bcbc57f1066b79f8']    # Pakistani Restaurant   

# Request the data
df_venues, query_list = GetVenuesByBox((fsw,fne), restaurant_codes)
                     
# Drop restaurants not in Georgia.
df_south_asia = df_venues[df_venues['state'].isin(['GA', 'Georgia'])]

......................................

In [25]:
# Create a list of categories we DON'T want. 
# List generated using this command: sorted(list(set(df_south_asia['categories'])))
# Categories to KEEP are commented OUT

bad_se = ['Arcade',
 'Bar',
 'Big Box Store',
 'Breakfast Spot',
 'Brewery',
# 'Chaat Place',
 'Chinese Restaurant',
 'Fast Food Restaurant',
# 'Food & Drink Shop',
 'Food Truck',
# 'Gourmet Shop',
 'Grocery Store',
 'Hotel',
# 'Indian Chinese Restaurant',
# 'Indian Restaurant',
# 'Indian Sweet Shop',
 'Italian Restaurant',
 'Mediterranean Restaurant',
# 'North Indian Restaurant',
# 'Pakistani Restaurant',
 'Pizza Place',
 'Sandwich Place',
# 'South Indian Restaurant',
 'Sports Bar',
# 'Tea Room',
 'Vegetarian / Vegan Restaurant']

# Drop all rows with a category we don't want
df_south_asia = df_south_asia[~df_south_asia['categories'].isin(bad_se)]

In [26]:
# Let's look at the number of restaurants by category
df_south_asia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name  lat  lng  state
categories                                      
Indian Restaurant           269  269  269    269
North Indian Restaurant       7    7    7      7
South Indian Restaurant       6    6    6      6
Indian Chinese Restaurant     5    5    5      5
Pakistani Restaurant          4    4    4      4
Chaat Place                   3    3    3      3
Food & Drink Shop             1    1    1      1
Gourmet Shop                  1    1    1      1
Indian Sweet Shop             1    1    1      1
Tea Room                      1    1    1      1

In [27]:
# Consolidate the categories into one - 'Indo-Pak Restaurant'

df_south_asia['categories'] = 'Indo-Pak Restaurant'
df_south_asia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name  lat  lng  state
categories                                
Indo-Pak Restaurant   298  298  298    298

Looks good. There may be some mis-labeled restaurants in this dataframe, but we'll look at that after we merge the dataframes together.

In [28]:
# Merge the dataframes
df_georgia = df_georgia.append(df_south_asia)
df_georgia.reset_index(drop=True, inplace=True)
print(df_georgia.shape)


(3342, 5)


In [29]:
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name   lat   lng  state
categories                                    
Chinese Restaurant     1096  1096  1096   1096
Asian Restaurant        731   731   731    731
Japanese Restaurant     724   724   724    724
Indo-Pak Restaurant     304   304   304    304
Korean Restaurant       216   216   216    216
Thai Restaurant         178   178   178    178
Vietnamese Restaurant    93    93    93     93

Did performing two queries and merging them introduce duplicates? Let's find out and remove if so.

In [30]:
print("Before dropping duplicates: ",df_georgia.shape[0])
df_georgia = df_georgia.drop_duplicates(keep='first')
print("After dropping duplicates: ",df_georgia.shape[0])

Before dropping duplicates:  3342
After dropping duplicates:  3337


Now that we have all the restaurants in one place, let's take a closer look at the data and fix some inaccuracies.

## Fixing Inaccurate Categories

After looking through the data, it's clear that the categorization of the restaurants can be improved. For example, there are plenty of restaurants in the general "Asian" category that have "Japanese" or "Chinese" in their names. There are also some restaurants that are just in the wrong category - again, Chinese restaurants that are categorized as Japanese, etc. Let's see if we can fix this programatically, and with a minimum of manual effort.

I extracted the names of the restaurants in each category, and did a word frequency analysis to find terms that are common and specific to each restaurant type. For example, Vietnamese restaurant names often have these terms:

    vietnamese_terms = ['Pho', 'Saigon', 'Viet', 'Banh Mi']
    
Let's see if we can use those terms to correct some assignment problems in this database.

In [31]:
# Define terms commonly found in restaurant names for each country of origin.

china_terms = ['China', 'Chinese', 'Wok', 'Hong Kong', 'Panda', 'Peking', 
               'Beijing', 'Great Wall']
japan_terms = ['Japanese', 'Tokyo', 'Japan', 'Osaka', 
               'Shogun', 'Fuji', 'Sumo', 'Ichiban', 'Kobe', 'Sakura', 'Ramen', 
               'Teriyaki', 'Ninja', 'Shabu']
korea_terms = ['Korea', 'Gogi']
thailand_terms = ['Thai', 'Bangkok']
vietnam_terms = ['Pho', 'Saigon', 'Viet', 'Banh Mi']
indopak_terms = ['India', 'Bombay', 'Biryani', 'Naan', 'Masala']

term_list = [['Chinese Restaurant',china_terms],
             ['Japanese Restaurant',japan_terms],
             ['Korean Restaurant', korea_terms],
             ['Thai Restaurant', thailand_terms],
             ['Vietnamese Restaurant', vietnam_terms],
             ['Indo-Pak Restaurant', indopak_terms]]

In [32]:
# Use keywords to reassign restaurants into more accurate categories
# Reset index to ensure it is unique and continuous
df_georgia.reset_index(drop=True, inplace = True)
# Iterate through the dataframe
for i in range(df_georgia.shape[0]):
    # For each row, see if any keyword appears in the restaurant name
    # If so, change the category
    for a_term in term_list:
        if any(x in df_georgia.loc[i,'name'] for x in a_term[1]):
            df_georgia.loc[i, 'categories'] = a_term[0]

Our data cleaning is complete. What does the cleaned database look like?

In [52]:
print(df_georgia.shape)
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False)

(3337, 13)


name   lat   lng  state  Asian  Chinese  Indo-Pak  \
categories                                                                 
Chinese Restaurant     1223  1223  1223   1223   1223     1223      1223   
Japanese Restaurant     800   800   800    800    800      800       800   
Asian Restaurant        476   476   476    476    476      476       476   
Indo-Pak Restaurant     295   295   295    295    295      295       295   
Korean Restaurant       217   217   217    217    217      217       217   
Thai Restaurant         204   204   204    204    204      204       204   
Vietnamese Restaurant   122   122   122    122    122      122       122   

                       Japanese  Korean  Thai  Vietnamese  cluster  
categories                                                          
Chinese Restaurant         1223    1223  1223        1223     1223  
Japanese Restaurant         800     800   800         800      800  
Asian Restaurant            476     476   476         476      476  
Indo-Pak Restaurant         295     295   295         295      295  
Korean Restaurant           217     217   217         217      217  
Thai Restaurant             204     204   204         204      204  
Vietnamese Restaurant       122     122   122         122      122

Now that we have our data, let's get an idea of where the restaurants are in the state.

# Visualization
Before we create sales territories from the data, let's see where our restaurants are located.


In [40]:
# Center the map on the middle of the bounding box defined above 
center = [(fsw[0]+fne[0])/2, (fsw[1]+fne[1])/2]
map_georgia = folium.Map(location=center, zoom_start=7, control_scale=True)   

heat_data = [[row['lat'],row['lng']] for index, row in df_georgia.iterrows()]

plugins.HeatMap(heat_data, radius=14).add_to(map_georgia)

# DELETE PRIOR
plugins.ScrollZoomToggler().add_to(map_georgia)
    
map_georgia

The heatmap shows that bulk of restaurants are in the greater Atlanta area, but it also seems to imply that the highest concentrations are outside the area. Further, I see different results depending on where I run this Jupyter notebook -- sometimes Columbus shows up as a major concentration, and other times it's Warner Robins or Savannah. Fortunately, Folium has another way to see thie information.

In [58]:
# Use cluster markers to show distribution of Asian restaurants in the state.
print(f'Number of Asian restaurants in Georgia: {df_georgia.shape[0]}')

# Center the map on the middle of the bounding box defined above 
center = [(fsw[0]+fne[0])/2, (fsw[1]+fne[1])/2]

map_georgia = folium.Map(location=center, zoom_start=7,  control_scale=True)

# Create Marker Cluster
all_cluster = plugins.MarkerCluster(control=False)
map_georgia.add_child(all_cluster)

sub_group_dict = {}  # For marker cluster sub-groups

# Create a list of categories, ordered by most-to-least popular
category_list = list(df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False).index)

# Create a sub-group for each type of restaurant
for category in category_list:
    sub_group_dict[category] = plugins.FeatureGroupSubGroup(all_cluster,category)
    map_georgia.add_child(sub_group_dict[category])

#DELETE SCROLLZOOMTOGGLER PRIOR TO SUBMITTING PROJECT
plugins.ScrollZoomToggler().add_to(map_georgia)

# Assign each restaurant to the appropriate sub-group
for restaurant in df_georgia.iterrows():
    folium.Marker(location = [restaurant[1]['lat'],restaurant[1]['lng']]
                  #, popup=restaurant[1]['name']  # Overloads system to add this
                 ).add_to(sub_group_dict[restaurant[1]['categories']])

# Display layer control
folium.LayerControl(collapsed=False).add_to(map_georgia)
map_georgia

Number of Asian restaurants in Georgia: 3337


This clarifies the picture. Of the roughly 3300 Asian restaurants in the state, over 2400 are in the greater Atlanta area, with a strong concentration in the communities to the Northeast along Interstate 85. The category checkboxes allow us to see how the distribution varies by type of restaurant. While all types are concentrated in the population centers around Atlanta, Chinese and Japanese restaurants are more evenly distributed around the state. They are also the most popular categories.

Now let's divide the potential customer base into sales territories for our restaurant supply company.

# Analysis

We will find our sales territories by using k means clustering. To do that, we must have numeric values for each category we use for the clustering analysis. We use one-hot encoding to convert the "categories" column to a group of columns for each possible value in "categories". 

In [34]:
# Create the dummy columns as a separate dataframe, 
# then check the sizes of the original and dummies dataframes.

df_georgia_dummies = pd.get_dummies(df_georgia['categories'])
print(df_georgia.shape)
print(df_georgia_dummies.shape)
df_georgia_dummies.head(10)

(3337, 5)
(3337, 7)


Asian Restaurant  Chinese Restaurant  Indo-Pak Restaurant  \
0                 1                   0                    0   
1                 1                   0                    0   
2                 0                   1                    0   
3                 0                   1                    0   
4                 0                   1                    0   
5                 0                   0                    0   
6                 0                   0                    0   
7                 1                   0                    0   
8                 1                   0                    0   
9                 0                   1                    0   

   Japanese Restaurant  Korean Restaurant  Thai Restaurant  \
0                    0                  0                0   
1                    0                  0                0   
2                    0                  0                0   
3                    0                  0                0   
4                    0                  0                0   
5                    1                  0                0   
6                    1                  0                0   
7                    0                  0                0   
8                    0                  0                0   
9                    0                  0                0   

   Vietnamese Restaurant  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
5                      0  
6                      0  
7                      0  
8                      0  
9                      0

In [35]:
# Change the column names to make them a bit easier to display

df_georgia_dummies.columns = ['Asian', 'Chinese', 'Indo-Pak',
       'Japanese', 'Korean', 'Thai',
       'Vietnamese']
df_georgia_dummies.head(10)

Asian  Chinese  Indo-Pak  Japanese  Korean  Thai  Vietnamese
0      1        0         0         0       0     0           0
1      1        0         0         0       0     0           0
2      0        1         0         0       0     0           0
3      0        1         0         0       0     0           0
4      0        1         0         0       0     0           0
5      0        0         0         1       0     0           0
6      0        0         0         1       0     0           0
7      1        0         0         0       0     0           0
8      1        0         0         0       0     0           0
9      0        1         0         0       0     0           0

In [36]:
# Now add the new columns to the original dataframe

df_georgia[list(df_georgia_dummies.columns)] = df_georgia_dummies

In [37]:
df_georgia.head()

name          categories        lat        lng state  \
0          Yuki Express    Asian Restaurant  30.903404 -84.537571    GA   
1                 Makan    Asian Restaurant  30.903800 -84.575470    GA   
2  Jin's Chinese Buffet  Chinese Restaurant  30.891642 -84.207543    GA   
3     Star China Buffet  Chinese Restaurant  30.902765 -84.554162    GA   
4             No1 China  Chinese Restaurant  30.886484 -84.205227    GA   

   Asian  Chinese  Indo-Pak  Japanese  Korean  Thai  Vietnamese  
0      1        0         0         0       0     0           0  
1      1        0         0         0       0     0           0  
2      0        1         0         0       0     0           0  
3      0        1         0         0       0     0           0  
4      0        1         0         0       0     0           0

In [39]:
# DELETE PRIOR TO PROJECT COMPLETION
# Now let's save this as a csv file.
file_contents = df_georgia.to_csv(index=False)
with(open('georgia.csv', 'w')) as file1:
    file1.write(file_contents)

In [44]:
# Set the number of clusters (equal to the number of salespeople). 
kclusters = 4

# df_georgia = df_georgia[df_georgia['categories'] != 'Indo-Pak Restaurant']

# georgia_clustering = df_georgia[['lat', 'lng']]

try:
    georgia_clustering = df_georgia.drop(['name','categories','state','cluster'], 1)
except:
    georgia_clustering = df_georgia.drop(['name','categories','state'], 1)

# Scale up/down latitude & longitude to adjust relative importance of driving time.  

distance_modifier = 0.5

georgia_clustering['lat'] *= distance_modifier    
georgia_clustering['lng'] *= distance_modifier     
    
    
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=27).fit(georgia_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:20] 

array([3, 3, 0, 0, 0, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 2, 2, 2, 2, 2],
      dtype=int32)

In [45]:
georgia_clustering['cluster']=kmeans.labels_
georgia_clustering.head()
df_georgia['cluster']=kmeans.labels_

In [46]:

# df_simple=df_georgia[['categories','cluster']]

# for restaurant in df_simple.iterrows():
#     restaurant[1][1] = color_dict[restaurant[1][1]]

# df_cross.columns[0]


# [0:4] = [color_dict[i] for i in list(df_cross.columns[0:4])]



# df_simple.groupby(['categories','cluster']).count()
# df_simple
# pd.pivot_table(df_simple, index=['categories'])

# df_cross

# .columns[0:3]
# [color_dict[i] for i in list(df_cross.columns[0:4])]

pd.crosstab(df_georgia.categories, df_georgia.cluster, #colnames = [0, 1, 2, 3, 'Total'],
                       margins=True, margins_name="Total")



cluster                   0    1    2     3  Total
categories                                        
Asian Restaurant          0    0   60   416    476
Chinese Restaurant     1018    0  205     0   1223
Indo-Pak Restaurant       0    0   20   275    295
Japanese Restaurant       0  691  109     0    800
Korean Restaurant         0    0   11   206    217
Thai Restaurant           0    0   26   178    204
Vietnamese Restaurant     0    0    8   114    122
Total                  1018  691  439  1189   3337

In [48]:
# Create a dictionary to control colors on map
color_dict = {
 0: 'yellow',
 1: 'red',
 2: 'blue',
 3: 'orange',
 4: 'cyan',
 5: 'pink',
 6: 'green'}


In [51]:
# Center the map on the middle of the bounding box defined above 
center = [(fsw[0]+fne[0])/2, (fsw[1]+fne[1])/2]
map_georgia = folium.Map(location=center, zoom_start=7, control_scale=True) 

plugins.ScrollZoomToggler().add_to(map_georgia)

# for restaurant in df_georgia[df_georgia['categories'].isin(['Indo-Pak Restaurant','Thai Restaurant', 'Korean Restaurant','Vietnamese Restaurant'])].iterrows():
#     folium.Circle(radius = 20, color = color_dict[restaurant[1]['cluster']], 
#                   location = [restaurant[1]['lat'],restaurant[1]['lng']]).add_to(map_georgia)

for restaurant in df_georgia.iterrows():
    folium.Circle(radius = 20, color = 'blue',# color_dict[restaurant[1]['cluster']], 
                  location = [restaurant[1]['lat'],restaurant[1]['lng']]).add_to(map_georgia)
map_georgia

cluster                  0    1    2    3   All
categories                                     
Asian Restaurant         0    0   45  425   470
Chinese Restaurant     990    0  205    0  1195
Japanese Restaurant      0  682  109    0   791
Korean Restaurant        0    0   11  204   215
Thai Restaurant          0    0   26  177   203
Vietnamese Restaurant    0    0    8  113   121
All                    990  682  404  919  2995

AttributeError: 'int' object has no attribute 'name'

In [0]:
x = 0
print(color_dict[x])
df_georgia[df_georgia['cluster'].isin([x])].groupby('categories').count().sort_values(by=['name'], ascending=False)

yellow


KeyError: 'cluster'

In [0]:
x = 1
print(color_dict[x])
df_georgia[df_georgia['cluster'].isin([x])].groupby('categories').count().sort_values(by=['name'], ascending=False)

NameError: name 'color_dict' is not defined

In [0]:
x = 2
print(color_dict[x])
df_georgia[df_georgia['cluster'].isin([x])].groupby('categories').count().sort_values(by=['name'], ascending=False)

blue


name  lat  lng  state  Asian  Chinese  Indo-Pak  Japanese  \
categories                                                                      
Chinese Restaurant   990  990  990    990    990      990       990       990   

                    Korean  Thai  Vietnamese  cluster  
categories                                             
Chinese Restaurant     990   990         990      990

In [0]:
x = 3
print(color_dict[x])
df_georgia[df_georgia['cluster'].isin([x])].groupby('categories').count().sort_values(by=['name'], ascending=False)

orange


name  lat  lng  state  Asian  Chinese  Indo-Pak  Japanese  \
categories                                                                    
Asian Restaurant   428  428  428    428    428      428       428       428   

                  Korean  Thai  Vietnamese  cluster  
categories                                           
Asian Restaurant     428   428         428      428

In [0]:
x = 4
print(color_dict[x])
df_georgia[df_georgia['cluster'].isin([x])].groupby('categories').count().sort_values(by=['name'], ascending=False)

cyan


name  lat  lng  state  Asian  Chinese  Indo-Pak  \
categories                                                               
Korean Restaurant       210  210  210    210    210      210       210   
Thai Restaurant         180  180  180    180    180      180       180   
Vietnamese Restaurant   113  113  113    113    113      113       113   

                       Japanese  Korean  Thai  Vietnamese  cluster  
categories                                                          
Korean Restaurant           210     210   210         210      210  
Thai Restaurant             180     180   180         180      180  
Vietnamese Restaurant       113     113   113         113      113

In [0]:
s = "Hello world how are you doing"
"world|howdy" in s

False

In [0]:
df_georgia[df_georgia['name'].str.contains('|'.join(term_list[0][1]))]['categories'] = term_list[0][0]

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [0]:
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name   lat   lng  state
categories                                    
Chinese Restaurant     1071  1071  1071   1071
Asian Restaurant        720   720   720    720
Japanese Restaurant     715   715   715    715
Indo-Pak Restaurant     297   297   297    297
Korean Restaurant       214   214   214    214
Thai Restaurant         177   177   177    177
Vietnamese Restaurant    92    92    92     92

In [0]:
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name   lat   lng  state  Asian  Chinese  Indo-Pak  \
categories                                                                 
Chinese Restaurant     1195  1195  1195   1195   1195     1195      1195   
Japanese Restaurant     791   791   791    791    791      791       791   
Asian Restaurant        470   470   470    470    470      470       470   
Indo-Pak Restaurant     293   293   293    293    293      293       293   
Korean Restaurant       215   215   215    215    215      215       215   
Thai Restaurant         203   203   203    203    203      203       203   
Vietnamese Restaurant   121   121   121    121    121      121       121   

                       Japanese  Korean  Thai  Vietnamese  
categories                                                 
Chinese Restaurant         1195    1195  1195        1195  
Japanese Restaurant         791     791   791         791  
Asian Restaurant            470     470   470         470  
Indo-Pak Restaurant         293     293   293         293  
Korean Restaurant           215     215   215         215  
Thai Restaurant             203     203   203         203  
Vietnamese Restaurant       121     121   121         121

In [0]:
list(set(df_georgia['categories']))

['Indo-Pak Restaurant',
 'Japanese Restaurant',
 'Korean Restaurant',
 'Chinese Restaurant',
 'Asian Restaurant',
 'Thai Restaurant',
 'Vietnamese Restaurant']

In [0]:
# THIS ROW DEMONSTRATES HOW TO FIND PARTIAL MATCHES.

df_whatsit = df_georgia[df_georgia['name'].str.contains('|'.join(thai_terms))]
df_whatsit.groupby('categories').count().sort_values(by=['name'], ascending=False)


name  lat  lng  state
categories                                
Thai Restaurant       145  145  145    145
Asian Restaurant       21   21   21     21
Japanese Restaurant     6    6    6      6
Chinese Restaurant      2    2    2      2

In [0]:
df_whatsit[df_whatsit['categories'].isin(['Chinese Restaurant'])]

name          categories        lat  \
1066  Lucky Buddha Chinese Thai And Sushi  Chinese Restaurant  33.782012   
1839             The Pearl Chinese & Thai  Chinese Restaurant  34.022653   

            lng state  
1066 -84.404881    GA  
1839 -84.270607    GA

In [0]:
for term in list(df_georgia['name'][df_georgia['categories'].isin(['Indo-Pak Restaurant'])]):
    print(term,end = " ")

Hsu's Gourmet Heart of India Restaurant Jade Dragon Carryout Dynasty Chinese Restaurant Paradise Biryani Pointe Simha Authentic Multi Cuisine Fine Dine Restaurant and Bar 

In [0]:
set(df_south_asia['categories'])

{'Chaat Place',
 'Food & Drink Shop',
 'Gourmet Shop',
 'Indian Chinese Restaurant',
 'Indian Restaurant',
 'Indian Sweet Shop',
 'Italian Restaurant',
 'North Indian Restaurant',
 'Pakistani Restaurant',
 'Pizza Place',
 'South Indian Restaurant',
 'Tea Room'}

In [0]:

sorted(list(set(df_georgia['categories'])))

['Asian Restaurant',
 'Chinese Restaurant',
 'Indo-Pak Restaurant',
 'Japanese Restaurant',
 'Korean Restaurant',
 'Thai Restaurant',
 'Vietnamese Restaurant']

In [0]:
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False)

name   lat   lng  state
categories                                    
Chinese Restaurant     1071  1071  1071   1071
Asian Restaurant        720   720   720    720
Japanese Restaurant     715   715   715    715
Korean Restaurant       214   214   214    214
Thai Restaurant         177   177   177    177
Vietnamese Restaurant    92    92    92     92
Indo-Pak Restaurant       6     6     6      6

In [0]:
a_dictionary = {'Indonesian Restaurant': 'a',
 'Poke Place': 'b',
 'Noodle House': 'c',
 'Ramen Restaurant': 'd',
 'Bubble Tea Shop': 'e',
 'Indian Restaurant': 'f',
 'Steakhouse': 'g'}

In [0]:
a_dictionary['Noodle House']

'c'

In [0]:
df_georgia[['categories','name']].groupby('categories').count()\
.sort_values(by='name', ascending=False)

name
categories                     
Chinese Restaurant         1044
Asian Restaurant            655
Japanese Restaurant         449
Sushi Restaurant            243
Korean Restaurant           211
Thai Restaurant             177
Vietnamese Restaurant        92
Noodle House                 32
Ramen Restaurant             14
Szechuan Restaurant           9
Taiwanese Restaurant          7
Malay Restaurant              5
Shabu-Shabu Restaurant        4
Supermarket                   4
Hotpot Restaurant             4
Indonesian Restaurant         4
Dim Sum Restaurant            4
Cantonese Restaurant          3
BBQ Joint                     3
Mongolian Restaurant          3
Indian Chinese Restaurant     3
Poke Place                    3
Sandwich Place                2
Steakhouse                    2
Dumpling Restaurant           2
Japanese Curry Restaurant     2
Indian Restaurant             2
Food Truck                    2
Karaoke Bar                   2
Filipino Restaurant           1
Middle Eastern Restaurant     1
Jiangsu Restaurant            1
Shanghai Restaurant           1
Soup Place                    1
Hot Dog Joint                 1
Garden                        1
Buffet                        1
Fried Chicken Joint           1
Taco Place                    1
Tea Room                      1
Bubble Tea Shop               1

# READ CSV FILES HERE

In [0]:
df_south_asia = pd.read_csv('https://raw.githubusercontent.com/JamesDCage/Final-Week-0/master/south_asian.csv')
df_south_asia.shape

(296, 6)

In [0]:
df_georgia = pd.read_csv('https://raw.githubusercontent.com/JamesDCage/Final-Week-0/master/georgia.csv')
df_georgia.shape

(3288, 12)

# BLANK LINES FOR TESTING

In [0]:
# Create a dictionary to control colors on map

color_dict = {'Asian Restaurant':'blue',
 'Chinese Restaurant':'red',
 'Indo-Pak Restaurant': 'orange',
 'Japanese Restaurant':'green',
 'Korean Restaurant': 'cyan',
 'Thai Restaurant': 'yellow',
 'Vietnamese Restaurant': 'pink'}


In [0]:
# Experiment in tryhing to put markers on map for restaurants. Too many markers, doesn't show up clearly. 

# Center the map on the middle of the bounding box defined above 
center = [(fsw[0]+fne[0])/2, (fsw[1]+fne[1])/2]# Visualize all queries on a map, with fill color indicating number of restaurants
map_georgia = folium.Map(location=center, zoom_start=7, control_scale=True) 

plugins.ScrollZoomToggler().add_to(map_georgia)

for restaurant in df_georgia[~df_georgia['categories'].isin(['Indo-Pak Restaurant','Thai Restaurant', 'Korean Restaurant','Vietnamese Restaurant'])].iterrows():
    folium.Circle(radius = 100, color = color_dict[restaurant[1]['categories']], 
                  opacity = 0.5, fill_color = color_dict[restaurant[1]['categories']], fill_opacity = 0.25,
                  location = [restaurant[1]['lat'],restaurant[1]['lng']]).add_to(map_georgia)
    
    
map_georgia

In [0]:
# SHOW DENSITY AS BOX COLOR
# Basically, nothing shows up but my neighborhood.

# Center the map on the middle of the bounding box defined above 
center = [(fsw[0]+fne[0])/2, (fsw[1]+fne[1])/2]

# Find the maximum number of venues in any square (to calculate fill opacity)
max_clients = max([query[1] for query in query_list])
min_box = min([(query[0][0][0]-query[0][1][0])*(query[0][0][1]-query[0][1][1]) for query in query_list])

# Visualize all queries on a map, with fill color indicating number of restaurants
map_georgia = folium.Map(location=center, zoom_start=7, control_scale=True)

plugins.ScrollZoomToggler().add_to(map_georgia)
    
for box in query_list:
    sw, ne = box[0][0], box[0][1]
    # Calculate area
    area = (ne[0]-sw[0])*(ne[1]-sw[1])
#     opacity = box[1]/max_clients   # Clients in this box / max across all boxes
    opacity = (box[1] * min_box)/(max_clients * area)
    folium.Rectangle([sw, ne], 
                     popup=str(box[1]), color="black", opacity=0.99, fill=True, 
                     fill_color='red', weight = 1, 
                     fill_opacity=(opacity)).add_to(map_georgia)
    
map_georgia

In [0]:
# CLUSTER MARKERS THAT AREN'T SEPARATD BY CATEGORY

# Use cluster markers to show distribution of Asian restaurants in the state.
print(f'Number of Asian restaurants in Georgia: {df_georgia.shape[0]}')

# Center the map on the middle of the bounding box defined above 
center = [(fsw[0]+fne[0])/2, (fsw[1]+fne[1])/2]

map_georgia = folium.Map(location=center, zoom_start=7, control_scale=True)   

#DELETE SCROLLZOOMTOGGLER PRIOR TO SUBMITTING PROJECT
plugins.ScrollZoomToggler().add_to(map_georgia)

all_cluster = plugins.MarkerCluster().add_to(map_georgia)

for restaurant in df_georgia.iterrows():
    folium.Marker(location = [restaurant[1]['lat'],restaurant[1]['lng']]
                  #, popup=restaurant[1]['name']
                 ).add_to(all_cluster)
    
map_georgia

# New Section

In [0]:
df_georgia.describe(include='all')

name          categories          lat          lng state  \
count            3288                3288  3288.000000  3288.000000  3288   
unique           2530                   7          NaN          NaN     2   
top     Panda Express  Chinese Restaurant          NaN          NaN    GA   
freq               55                1195          NaN          NaN  3199   
mean              NaN                 NaN    33.541198   -83.917925   NaN   
std               NaN                 NaN     0.808775     0.954345   NaN   
min               NaN                 NaN    30.681651   -85.512472   NaN   
25%               NaN                 NaN    33.471594   -84.385496   NaN   
50%               NaN                 NaN    33.881895   -84.212470   NaN   
75%               NaN                 NaN    33.993296   -83.958817   NaN   
max               NaN                 NaN    34.971943   -80.848510   NaN   

              Asian      Chinese     Indo-Pak     Japanese       Korean  \
count   3288.000000  3288.000000  3288.000000  3288.000000  3288.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean       0.142944     0.363443     0.089112     0.240572     0.065389   
std        0.350069     0.481064     0.284949     0.427496     0.247249   
min        0.000000     0.000000     0.000000     0.000000     0.000000   
25%        0.000000     0.000000     0.000000     0.000000     0.000000   
50%        0.000000     0.000000     0.000000     0.000000     0.000000   
75%        0.000000     1.000000     0.000000     0.000000     0.000000   
max        1.000000     1.000000     1.000000     1.000000     1.000000   

               Thai  Vietnamese  
count   3288.000000   3288.0000  
unique          NaN         NaN  
top             NaN         NaN  
freq            NaN         NaN  
mean       0.061740      0.0368  
std        0.240719      0.1883  
min        0.000000      0.0000  
25%        0.000000      0.0000  
50%        0.000000      0.0000  
75%        0.000000      0.0000  
max        1.000000      1.0000

Empty DataFrame
Columns: [name, categories, lat, lng, state, s_asia]
Index: []

In [0]:
df_georgia[df_georgia['categories'].isin(['Soup Place'])].sort_values(by='categories')

name  categories        lat        lng state
2032  Juicy & Tasty  Soup Place  33.950916 -84.140539    GA

In [0]:
df_georgia.describe(include='all')

name         categories         lat         lng state
count             322                322  322.000000  322.000000   322
unique            296                 26         NaN         NaN     2
top     Chinese Dhaba  Indian Restaurant         NaN         NaN    GA
freq                4                265         NaN         NaN   313
mean              NaN                NaN   33.749211  -84.111439   NaN
std               NaN                NaN    0.601438    0.640293   NaN
min               NaN                NaN   30.835180  -85.258985   NaN
25%               NaN                NaN   33.796777  -84.349300   NaN
50%               NaN                NaN   33.903164  -84.271429   NaN
75%               NaN                NaN   34.035847  -84.164537   NaN
max               NaN                NaN   34.959035  -81.090310   NaN

In [0]:
set(df_georgia['categories'])

{'Arcade',
 'Bar',
 'Big Box Store',
 'Breakfast Spot',
 'Brewery',
 'Chaat Place',
 'Chinese Restaurant',
 'Fast Food Restaurant',
 'Food & Drink Shop',
 'Food Truck',
 'Gourmet Shop',
 'Grocery Store',
 'Hotel',
 'Indian Chinese Restaurant',
 'Indian Restaurant',
 'Indian Sweet Shop',
 'Italian Restaurant',
 'Mediterranean Restaurant',
 'North Indian Restaurant',
 'Pakistani Restaurant',
 'Pizza Place',
 'Sandwich Place',
 'South Indian Restaurant',
 'Sports Bar',
 'Tea Room',
 'Vegetarian / Vegan Restaurant'}

In [0]:
df_georgia[df_georgia['categories'].isin(['Ramen Restaurant'])]

name        categories        lat        lng  \
329                          Sumo  Ramen Restaurant  31.961258 -83.759125   
507                         Akedo  Ramen Restaurant  32.052150 -81.103480   
1091                      Ton Ton  Ramen Restaurant  33.773323 -84.365836   
1092                Ramen Station  Ramen Restaurant  33.746380 -84.370605   
1226                  Taiyo Ramen  Ramen Restaurant  33.776722 -84.296559   
1227  Pao Pao Ramen Factory & Bar  Ramen Restaurant  33.815229 -84.311979   
1429               Momo San Ramen  Ramen Restaurant  33.954304 -84.551480   
1626              Jinya Ramen Bar  Ramen Restaurant  33.856618 -84.382846   
1761                     Yebisuya  Ramen Restaurant  33.908140 -84.287104   
1991                  Kumai Ramen  Ramen Restaurant  33.951736 -84.141161   
2407             Hajime Ramen Bar  Ramen Restaurant  34.059297 -84.235073   
2408                       Hajime  Ramen Restaurant  34.059430 -84.234890   
2680                      CupPoki  Ramen Restaurant  34.064748 -83.984139   
2681        Pao Pao Ramen Factory  Ramen Restaurant  34.069023 -83.986660   

     state  
329     GA  
507     GA  
1091    GA  
1092    GA  
1226    GA  
1227    GA  
1429    GA  
1626    GA  
1761    GA  
1991    GA  
2407    GA  
2408    GA  
2680    GA  
2681    GA

In [0]:
df_georgia.groupby('categories').count().sort_values(by=['name'], ascending=False).tail(19)

name  lat  lng  state
categories                                          
Big Box Store                     2    2    2      2
Food Truck                        2    2    2      2
Italian Restaurant                1    1    1      1
Tea Room                          1    1    1      1
Sports Bar                        1    1    1      1
Sandwich Place                    1    1    1      1
Pizza Place                       1    1    1      1
Mediterranean Restaurant          1    1    1      1
Arcade                            1    1    1      1
Indian Sweet Shop                 1    1    1      1
Bar                               1    1    1      1
Hotel                             1    1    1      1
Gourmet Shop                      1    1    1      1
Food & Drink Shop                 1    1    1      1
Fast Food Restaurant              1    1    1      1
Chinese Restaurant                1    1    1      1
Brewery                           1    1    1      1
Breakfast Spot                    1    1    1      1
Vegetarian / Vegan Restaurant     1    1    1      1

In [0]:
df_georgia.shape

(3339, 12)